adding spec md file (#2)
MatthewMachuca/FlexboxApiSPEC: flexboxAPIspecV2.md

Flexbox Layout for WinUI 
===

# Background

Flexbox Layout is designed to provide an efficient way to layout, align, and distribute space among items within a container, specifically when the size of the device is either unknown or dynamic.

Flexbox is a widely used layout schematic in web development, it allows developers to quickly create complex layouts with a standardized system. By introducing Flexbox to WinUI we can increase the layout choices developers have and cater to those who desire to use this popular layout in their WinUI applications.

More on Flexbox available here: https://css-tricks.com/snippets/css/a-guide-to-flexbox/#aa-flexbox-properties

# Remarks

WinUI Offers several Layout options, such as StackLayout and Flowlayout. Usage of Flexbox Layout will behave similarly and interact by wrapping your desired elements in a Layout Control.

#### Flex Containers vs Flex Items

> Flex Containers are elements that hold and specify the arangement of Flex Items. Containers have the following accepted properties:

Reviewer: Still confused about what containers & items are. Is there a <FlexContainer> type? Or is the first child of the Flexbox a container and its children or items? Probably just needs a line or two of descriptive text and some example markup.

| Property | 
|-|
| Direction |
Wrap|
JustifyContent|
AlignItems|
AlignContent|

Flex Items are elements that are children of a Flex Container. Items have the following accepted properties:
| Property | 
|-|
Order|
Grow|
Shrink|
Basis|
AlignSelf|
Unit|

#### Traditional CSS Behavior vs Flexbox in WinUI

> Several behaviors may deviate from the exact behavior found in CSS Flexbox, to highlight one in XAML, items do not overflow their containers, instead they will be omitted from the layout.

Reviewer: Can we expand on this? More detail?

# API Pages

_(Each of the following L2 sections correspond to a page that will be on docs.microsoft.com)_

## Flex Direction

The flex-direction property specifies the direction of the flexible items.
Note: If the element is not a flexible item, the flex-direction property has no effect.

```xml
<controls:LayoutPanel>
    <controls:LayoutPanel.Layout>
        <controls:FlexboxLayout Direction="Row"/>
    </controls:LayoutPanel.Layout>

>             <Grid>

Reviewer: To make this easier to read, leave out the Grids and just do the TextBlocks

Author: The grids in the examples are used to provide a way to apply styling. Since there is no "background color" property for Textblocks

                <TextBlock>3</TextBlock>
            </Grid>
            <Grid>
                <TextBlock>4</TextBlock>
            </Grid>
</controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexDirectionRow.PNG)

##  Flex Direction Accepted Values

| Value | Description |
|-|-|
| Row (Default) | The flexible items are displayed horizontally, as a row |
| RowReverse | Same behavior as Row, but elemnts are shown in reverse |
| Column | The flexible items are displayed vertically, as a column |
| ColumnReverse | Same behavior as Column, but elemnts are shown in reverse |

# API Details

```c# (but really MIDL3)

runtimeclass FlexboxLayout;

    enum FlexboxDirection
        {
            Row = 0,
            RowReverse = 1,
            Column = 2,
            ColumnReverse = 3
        }
```

## Flex Wrap

The flex-wrap property specifies whether the flexible items should wrap or not.

Note: If the elements are not flexible items, the flex-wrap property has no effect.

```xml

<controls:LayoutPanel MaxWidth="500">
    <controls:LayoutPanel.Layout>
        <controls:FlexboxLayout Direction="Row" Wrap="Wrap"/>
    </controls:LayoutPanel.Layout>

>             <Grid MinWidth="10">

Reviewer: Again can leave out the Grids here. I'd make the content of the TextBlock be the value of the MinWidth, just to make the picture even easier to read (it's already good).

Author: I think I would need to manually re-do all of the example photos as they have the same styling scheme, which requires the grid for a background color

    be duplicated here.

    Examples:
    * New class (RadialGradientBrush):
      https://github.com/microsoft/microsoft-ui-xaml-specs/blob/master/active/RadialGradientBrush/RadialGradientBrush.md
    * New member on an existing class (UIElement.ProtectedCursor):
      https://github.com/microsoft/microsoft-ui-xaml-specs/blob/master/active/UIElement/ElementCursor.md

    Style guide:
    * Use second person; speak to the developer who will be learning/using this API.
    (For example "you use this to..." rather than "the developer uses this to...")
    * Use hard returns to keep the page width within ~100 columns.
    (Otherwise it's more difficult to leave comments in a GitHub PR.)
    * Talk about an API's behavior, not its implementation.
    (Speak to the developer using this API, not to the team implementing it.)
    * A picture is worth a thousand words.
    * An example is worth a million words.
    * Keep examples realistic but simple; don't add unrelated complications.
    (An example that passes a stream needn't show the process of launching the File-Open dialog.)

-->

Flexbox Layout for WinUI 
===

> # Background

Reviewer: Open issue to call out: Should we have a Flexbox Panel (in addition to the FlexboxLayout)?

> <!-- TEMPLATE

Reviewer 2: Is it standard for these comment sections to be left in the file or should we remove these?

Reviewer: Nice to remove

    using an existing related API.

    For a simple example see the spec for the UIElement.ProtectedCursor property
    (https://github.com/microsoft/microsoft-ui-xaml-specs/blob/master/active/UIElement/ElementCursor.md)
    which has some of the thinking about how this Xaml API relates to existing
    Composition and WPF APIs. This is interesting background both for the current reader
    and the future reader trying to understand why we designed it this way,
    but not the kind of information
    that would land on DMC.
-->

Flexbox Layout is designed to provide an efficient way to layout, align, and distribute space among items within a container, specifically when the size of the device is either unknown or dynamic.

Flexbox is a widely used layout schematic in web development, it allows developers to quickly create complex layouts with a standardized system. By introducing Flexbox to WinUI we can increase the layout choices developers have and cater to those who desire to use this popular layout in their WinUI applications.

More on Flexbox available here: https://css-tricks.com/snippets/css/a-guide-to-flexbox/#aa-flexbox-properties

> # Remarks

Reviewer: The content here is fine, there's just no "Remarks" section in the template. So keep the sentence below about layout options, but remove the rest, and let it be part of Background.

<!-- TEMPLATE
    (Optional)

    All APIs have a page on DMC, some APIs or groups of APIs have an additional high level,
    conceptual page (called a "how-to" page). This section can be used for that content.

    For example, there are several Xaml controls for different forms of text input,
    each with an API page, and then there's also a conceptual page that
    discusses them collectively
    (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/text-controls).

    Another way to use this section is as a draft of a blog post that introduces the new feature.

    Sometimes it's difficult to decide if text belongs on a how-to page or an API page.
    It's not important to make a final decision on that in this spec; we can always
    adjust it when copying to DMC.
-->

WinUI Offers several Layout options, such as StackLayout and Flowlayout. Usage of Flexbox Layout will behave similarly and interact by wrapping your desired elements in a Layout Control.

> #### Flex Containers vs Flex Items

Reviewer: ```suggestion
# Remarks
## Flex Containers vs Flex Items
```

This and the part about CSS below looks like it wants to be part of the "Conceptual pages" section in the spec template. Good idea and good topic. 

There should be some more overview content ahead of that generally describing what Flex is (target audience being Xaml developers; this section will end up on docs.microsoft.com). The goal is to enable someone to learn enough that they can comprehend what's being said in the rest of the doc.

Flex Containers are elements that hold and specify the arangement of Flex Items. Containers have the following accepted properties:
| Property | 
|-|
| Direction |
Wrap|
JustifyContent|
AlignItems|
AlignContent|

Flex Items are elements that are children of a Flex Container. Items have the following accepted properties:
| Property | 
|-|

> order|

Reviewer 2: nit: make uppercase

Grow|
Shrink|
Basis|
AlignSelf|
Unit|

#### Traditional CSS Behavior vs Flexbox in WinUI

Several behaviors may deviate from the exact behavior found in CSS Flexbox, to highlight one in XAML, items do not overflow their containers, instead they will be omitted from the layout.

> # API Pages

Reviewer: There should be one API Pages section with all the APIs, then one API Details section with the IDL.

    create a section for it.
  * If its purpose and usage is fully explained by brief description, either
      put it in a table in the "Other [class] members" section
      put it with /// comments in the IDL section

  Create an L2 section here for each API that needs more description or examples.
  For a new class with members, the members should go in their own L2 section.

  Example layout
    ## MyClass
    ## MyClass.Member1
    ## MyClass.Member2
    ## Other MyClass members
    ## MyOtherClass
    ## ...

  Notes:
  * The first line of each of these sections should become that first line on the DMC page,
    which then becomes the description you see in Intellisense.
  * Each page can have description, examples, and remarks.
    Remarks are where the documentation calls out special considerations that the developer
    should be aware of.
  * It can be helpful at the top of an API page (or after the Intellisense text) to add the
    API signature in C#
  * Add a "_Spec note: ..._" to add a note that's useful in this spec but shouldn't go to DMC.
  * Show _examples_, not _samples_; an example is a snippet, a sample is a full working app.

-->

> ## Flex Direction

Reviewer: ```suggestion
## FlexDirection enum
```

Each of these L2 sections in the API Pages should be an API name. We don't need to have a section for every API though because some are obvious. For example once we describe FlexDirection here we don't need to also have a section for the FlexDirection property; just having that property in the IDL is good enough.

> The flex-direction property specifies the direction of the flexible items.

Reviewer 2: We might want to sub out the CSS syntax "flex-direction" for what the winUI syntax will be "Direction" in each of these api cases.

> The flex-direction property specifies the direction of the flexible items.

Reviewer: ```suggestion
The FlexDirection specifies the direction of the flexible items.
```

> Note: If the element is not a flexible item, the flex-direction property has no effect.

Reviewer: ```suggestion
Note: If the element is not a flexible item, the FlexDirection has no effect.
```

> 

Reviewer: Words explaining what's going on here, e.g.:

The following example shows a Flexbox laying out its children in rows.

```xml
<controls:LayoutPanel>
    <controls:LayoutPanel.Layout>
        <controls:FlexboxLayout Direction="Row"/>
    </controls:LayoutPanel.Layout>
            <Grid>
                    <TextBlock>1</TextBlock>
            </Grid>
            <Grid>
                <TextBlock>2</TextBlock>
            </Grid>
            <Grid>
                  <TextBlock>3</TextBlock>
            </Grid>
            <Grid>
                  <TextBlock>4</TextBlock>

>             </Grid>

Reviewer 3: Fix indentation

</controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexDirectionRow.PNG)

> ##  Flex Direction Accepted Values

Reviewer: Should be an L3 heading ("###") (or OK to leave it out too).

| Value | Description |
|-|-|
| Row (Default) | The flexible items are displayed horizontally, as a row |

> | RowReverse | Same behavior as Row, but elemnts are shown in reverse |

Reviewer: ```suggestion
| RowReverse | Same behavior as Row, but elements are shown in reverse |
```

| Column | The flexible items are displayed vertically, as a column |
| ColumnReverse | Same behavior as Column, but elemnts are shown in reverse |

# API Details

```c# (but really MIDL3)

runtimeclass FlexboxLayout;

>     enum FlexboxDirection

Reviewer 2: In the code snippet "Direction" is the property name, in the enum it is "FlexboxDirection". Which name is going to be used? Same for examples below.

Reviewer: It's OK for the property name to be shorter than the type name, since member names are always in the context of a class. So I think it looks good.

        {
            Row = 0,
            RowReverse = 1,
            Column = 2,
            ColumnReverse = 3
        }
```

> ## Flex Wrap

Reviewer: ```suggestion
## FlexWrap enum
```

> The flex-wrap property specifies whether the flexible items should wrap or not.

Reviewer: ```suggestion
Specifies whether the flexible items should wrap or not.
```

> Note: If the elements are not flexible items, the flex-wrap property has no effect.

Reviewer: ```suggestion
Note: If the elements are not flexible items, FlexWrap has no effect.
```

```xml

<controls:LayoutPanel MaxWidth="500">
    <controls:LayoutPanel.Layout>
        <controls:FlexboxLayout Direction="Row" Wrap="Wrap"/>
    </controls:LayoutPanel.Layout>
            <Grid MinWidth="10">
                <TextBlock>1</TextBlock>
            </Grid>
            <Grid MinWidth="20">
                <TextBlock>2</TextBlock>
            </Grid>
            <Grid MinWidth="100">
                <TextBlock>3</TextBlock>
            </Grid>
            <Grid MinWidth="120">
                <TextBlock>4</TextBlock>
            </Grid>
            <Grid MinWidth="120">
                <TextBlock>5</TextBlock>
            </Grid>
            <Grid MinWidth="60">
                <TextBlock>6</TextBlock>
            </Grid>
            <Grid MinWidth="60">
                <TextBlock>7</TextBlock>
            </Grid>
            <Grid MinWidth="100">
                <TextBlock>8</TextBlock>
            </Grid>

>         </controls:LayoutPanel>

Reviewer 3: Fix indentation

```
### Output of the above code (styling omitted)
![image](./FlexWrap.PNG)

##  Flex Wrap Accepted Values

| Value | Description |
|-|-|
| NoWrap (Default) | Default value. Specifies that the flexible items will not wrap |
| Wrap | Specifies that the flexible items will wrap if necessary	|
| WrapReverse | Specifies that the flexible items will wrap, if necessary, in reverse orde |

# API Details

```c# (but really MIDL3)

> runtimeclass FlexboxLayout;

Reviewer: Don't need this line

    enum FlexboxWrap
        {
            NoWrap = 0,
            Wrap = 1,
            WrapReverse = 2,
        }
```

## Flex JustifyContent

The justify-content property aligns the flexible container's items when the items do not use all available space on the main-axis (horizontally).

```xml

<controls:LayoutPanel>
    <controls:LayoutPanel.Layout>
        <controls:FlexboxLayout JustifyContent="Start"/>
    </controls:LayoutPanel.Layout>
            <Grid MinWidth="120"/>
            <Grid MinWidth="60"/>
            <Grid MinWidth="120"/>
</controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexJustifyContentStart.PNG)

##  Flex JustifyContent Accepted Values

| Value | Description |
|-|-|
| Start (Default) | Items are positioned at the beginning of the container |
| End | Items are positioned at the end of the container|
| Center | Items are positioned in the center of the container |
| SpaceBetween | Items will have space between them |

> | SpaceAround | Items are positioned at the end of the container|

Reviewer: This description doesn't match the name to me, is it written correctly? Might need more examples.

| SpaceEvenly | Items will have space before, between, and after them	 |

# API Details

```c# (but really MIDL3)

runtimeclass FlexboxLayout;

    enum FlexboxJustifyContent
        {
            Start = 0,
            End = 1,
            Center = 2,
            SpaceBetween = 3,
            SpaceAround = 4,
            SpaceEvenly = 5
        }
```

## Flex AlignItems

> The align-items property specifies the default alignment for items inside the flexible container.

Reviewer 3: `align-items` should be renamed to map to the XAML property name.
But do recommend linking to the css spec for each (I believe I provided a link in your previous doc).
Same comment for other similar properties in this review.

```xml

<controls:LayoutPanel>
            <controls:LayoutPanel.Layout>
                <controls:FlexboxLayout Direction="Row" AlignItems="Start"/>
            </controls:LayoutPanel.Layout>
            <Grid MinHeight="100"/>
            <Grid MinHeight="120"/>
            <Grid MinHeight="60"/>
            <Grid MinHeight="100"/>
        </controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexAlignItemsStart.PNG)

##  Flex AlignItems Accepted Values

| Value | Description |
|-|-|
| Stretch (Default) |Items are stretched to fit the container |
| Center | Items are positioned at the center of the container	|
| Start | Items are positioned at the beginning of the container|
| End | Items are positioned at the end of the container|

# API Details

```c# (but really MIDL3)

runtimeclass FlexboxLayout;

    enum FlexboxAlignItems
        {
            Start = 0,
            End = 1,
            Center = 2,
            Stretch = 3
        }
```

> 

Reviewer 2: nit: extra space

            <Grid controls:FlexboxLayout.Order="5">
                <TextBlock>5</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Order="-1">
                <TextBlock>-1</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Order="2">
                <TextBlock>2</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Order="1">
                <TextBlock>1</TextBlock>
            </Grid>
        </controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexOrder.PNG)

##  Flex Order Accepted Values

| Value | Description |
|-|-|
| Number (Default Value = 0) | Specifies the order for the flexible item |

# API Details

```c# (but really MIDL3)

unsealed runtimeclass FlexboxLayout : NonVirtualizingLayout
    
    {
        static Windows.UI.Xaml.DependencyProperty OrderProperty{ get; };
        static void SetOrder(Windows.UI.Xaml.UIElement element, Int32 value);
        static Int32 GetOrder(Windows.UI.Xaml.UIElement element);
    }
```

## Flex Grow

The flex-grow property specifies how much the item will grow relative to the rest of the flexible items inside the same container.

> Note: If the element is not a flexible item, the flex-grow property has no effect.

Reviewer: What does "flexible item" mean?

```xml
 <controls:LayoutPanel MinWidth="500">
            <controls:LayoutPanel.Layout>
                <controls:FlexboxLayout/>
            </controls:LayoutPanel.Layout>
            <Grid controls:FlexboxLayout.Grow="0" controls:FlexboxLayout.Basis="100" >
                <TextBlock>0</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Grow="1" controls:FlexboxLayout.Basis="100">
                <TextBlock>1</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Grow="0" controls:FlexboxLayout.Basis="100">
                <TextBlock>0</TextBlock>
            </Grid>
        </controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexGrow.PNG)

##  Flex Grow Accepted Values

| Value | Description |
|-|-|

> | Number (Default Value = 0) | Specifies how much the item will grow relative to the rest of the flexible items. |

Reviewer 4: this (and others) aren't numbers, but a struct comprised of a number and a `FlexboxUnit`.
Need to define this struct in this doc too.

Author: Could you give me an example of what this might look like?

Reviewer 4: You would first define something like:
```cs
struct FlexboxMeasurement
{
    Double Value;
    FlexboxUnit Unit;
}
```

then you would use that type in here, so the user can specify "20%" and that gets converted to a `FlexboxMeasurement` (name pending) with Value=20, and Unit=Percentage

# API Details

```c# (but really MIDL3)

unsealed runtimeclass FlexboxLayout : NonVirtualizingLayout
    
    {
        static Windows.UI.Xaml.DependencyProperty GrowProperty{ get; };
        static void SetGrow(Windows.UI.Xaml.UIElement element, Double value);

>         static Double GetGrow(Windows.UI.Xaml.UIElement element);

Reviewer 4: we had a discussion about float/Single vs. double/Double
In general XAML properties of "number" types are Double, so maybe it would make sense to keep all these IDL properties as Double, and then update the code to not use float and use double instead.

If that doesn't work for some reason, then using Single would be how you represent C++ "float"

Reviewer: Yes, default for numbers in Windows API is int or double

Author: Yes I did try changing everything to Single in the IDL file, but it ended up throwing errors and I left as is. Within the code changing everything to double is definitely an option

    }
```

## Flex Shrink

The flex-shrink property specifies how the item will shrink relative to the rest of the flexible items inside the same container.

> Note: If the element is not a flexible item, the flex-shrink property has no effect.

Reviewer 3: What does that mean?

```xml
<controls:LayoutPanel Width="500">
    <controls:LayoutPanel.Layout>
          <controls:FlexboxLayout/>
    </controls:LayoutPanel.Layout>
            <Grid controls:FlexboxLayout.Shrink="1" Width=" 250">
                <TextBlock>1</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Shrink="1" Width=" 250">
                <TextBlock>1</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Shrink="1" Width=" 250">
                <TextBlock>1</TextBlock>
            </Grid>
</controls:LayoutPanel>
```

### Output of the above code (styling omitted)

todo

##  Flex Shrink Accepted Values

| Value | Description |
|-|-|
| Number (Default Value = 1) | Specifies how much the item will shrink relative to the rest of the flexible items. |

# API Details

```c# (but really MIDL3)

unsealed runtimeclass FlexboxLayout : NonVirtualizingLayout
    
    {
         static Windows.UI.Xaml.DependencyProperty ShrinkProperty{ get; };
         static void SetShrink(Windows.UI.Xaml.UIElement element, Double value);
         static Double GetShrink(Windows.UI.Xaml.UIElement element);
    }
```

> ## Flex Basis

Reviewer: This could use some more explanation. Might be enough though to just add the explanatory text on the code sample. 

There should be an explicit mention that this is based on Basis and Unit.

# API Details

```c# (but really MIDL3)

runtimeclass FlexboxLayout;

    enum FlexboxAlignSelf
        {
            Auto = 0,
            Start = 1,
            End = 2,
            Center = 3,
            Stretch = 4,
        }
```

## Flexbox Unit

The flex-unit property will define the unit of measurement to be used by the flex-basis property.

```xml
        <controls:LayoutPanel MinWidth="500">
            <controls:LayoutPanel.Layout>
                <controls:FlexboxLayout/>
            </controls:LayoutPanel.Layout>
            <Grid controls:FlexboxLayout.Basis="50" Unit="px">
                <TextBlock>50</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Basis="100" Unit="px">
                <TextBlock>100</TextBlock>
            </Grid>
            <Grid controls:FlexboxLayout.Basis="150" Unit="px">
                <TextBlock>150</TextBlock>
            </Grid>
        </controls:LayoutPanel>
```

### Output of the above code (styling omitted)
![image](./FlexBasis.PNG)

##  Flex AlignSelf Accepted Values

| Value | Description |
|-|-|
| px (Default) |The unit of measurment input into the flex-basis property will be calculated in pixels |

> | % | The element's size will be calculated as a relative percentage of its container|

Reviewer 3: `%` won't work as a value for this. Will need to be `Pixels` and `Percent`.

# API Details

> 

Reviewer 2: Could we add some more information here on FlexboxUnit? i.e. background of why this new property is needed

```c# (but really MIDL3)

runtimeclass FlexboxLayout;

    enum FlexboxUnit
        {

>             px = 0,

Reviewer 4: Pixels

>             % = 1,

Reviewer 4: Percent, or Percentage

Reviewer: +1 to Percent

        }
```

# Appendix

> 

Reviewer 4: two open issues:
- *Layout*: how do flexbox layout properties interact (or don't) with XAML layout properties, e.g. `HorizontalAlignment` exists in both
- *Markup*: we'll likely need work in the parser (and some sort of `FlexboxUnitHelper` type) so that the xaml compiler can understand map:
  - "30"  --> { 30, FlexboxUnit::Pixels }
  - "30px" --> { 30, FlexboxUnit::Pixels }
  - "30%" --> { 30, FlexboxUnit::Percent }
